Develop two types of MLE for LM

sheny2 opened the pull request.

Wenxuan-Lu commented on tests/testthat/test-gradient-calculation.R:
In testthat, it is recommended to have one test case per function. Instead of separating test cases with comments, please organize them into separate test_that() function calls.

Wenxuan-Lu commented on tests/testthat/test-are-all-close.R:
In testthat, it is recommended to have one test case per function. Instead of separating test cases with comments, please organize them into separate test_that() function calls.

Wenxuan-Lu commented on tests/testthat/test-gradient-calculation.R:
-1pt; Avoid repeatedly defining the same function. Please do not redefine the log likelihood and gradient functions. For approx_grad(), move it into tests/testthat/helper.R as it may be needed to test other gradient functions when more implementations are done.

Wenxuan-Lu commented on R/hiper_glm.R:
Great job ensuring the validity of input parameters.

Wenxuan-Lu commented on R/hiper_glm.R:
It is recommended to have more but shorter functions in separate files for clarity and easier debugging. Please move neg_log_likelihood() and grad_neg_log_likelihood() into a separate file, such as R/likelihood_calculation.R. Also consider moving the actual implementation of pseudo-inverse and the codes calling BFGS to separated functions in a separate file, such as model_fitting.R. Though it does not seem to be necessary for the current implementation, this will make the code more modular and flexible, especially for developing a generic function that can accommodate various models and fitting methods.

Wenxuan-Lu commented on R/hiper_glm.R:
```diff
@@ -1,8 +1,81 @@
+    )
+
+    # Extract the estimated coefficients
+    coef_est <- as.matrix(optim_result$par, ncol = 1)
```
-1 pt; Please check whether BFGS has converged before returning the result. If it has not converged, provide a warning message to alert the user.

Wenxuan-Lu commented:
.Rproj.user is usually put in the .gitignore file because it contains user-specific settings and temporary files that do not need to be shared across a team.

Wenxuan-Lu reviewed (commented):
Great job! I enjoyed reviewing your code. Remember that you can earn back all points by addressing these comments. Please address both point-deducting and non-point-deducting comments. I will update this comment with the final score when HW3 is due.

sheny2 commented:
> Great job! I enjoyed reviewing your code. Remember that you can earn back all points by addressing these comments. Please address both point-deducting and non-point-deducting comments. I will update this comment with the final score when HW3 is due.

Hi Wenxuan, thank you for your comments. Just want to let you know that the merged develop branch has addressed all feedback that you provided. Plesae review if needed. 
I will continue working on logistic settings under `logit-model` branch.

Wenxuan-Lu commented:
> > Great job! I enjoyed reviewing your code. Remember that you can earn back all points by addressing these comments. Please address both point-deducting and non-point-deducting comments. I will update this comment with the final score when HW3 is due.
> 
> Hi Wenxuan, thank you for your comments. Just want to let you know that the merged develop branch has addressed all feedback that you provided. Plesae review if needed. I will continue working on logistic settings under `logit-model` branch.

Hi Yicheng, one more comment for the code. You don't need to copy the likelihood functions to tests/testthat/helper.R. Functions defined in the R folder can be called directly from the tests.

sheny2 commented:
> > > Great job! I enjoyed reviewing your code. Remember that you can earn back all points by addressing these comments. Please address both point-deducting and non-point-deducting comments. I will update this comment with the final score when HW3 is due.
> 
> > 
> 
> > Hi Wenxuan, thank you for your comments. Just want to let you know that the merged develop branch has addressed all feedback that you provided. Plesae review if needed. I will continue working on logistic settings under `logit-model` branch.
> 
> 
> 
> Hi Yicheng, one more comment for the code. You don't need to copy the likelihood functions to tests/testthat/helper.R. Functions defined in the R folder can be called directly from the tests.


Thank you for pointing that out! Yes I noticed that redundancy and have cleaned the extra codes in the latest the logic model branch.
